src/jtag/drivers/ftdi_cjtag.c

```diff
@@ -0,0 +1,787 @@
+
+	pucBuffer[ucCount++] = 0x2A;
+	pucBuffer[ucCount++] = 0x00;
+
+	pucBuffer[ucCount++] = 0x80;
+	if (tms_and_tdi == (TMS_0 | TDI_0))
+		pucBuffer[ucCount++] = 0xA0;
+	else if (tms_and_tdi == (TMS_0 | TDI_1))
+		pucBuffer[ucCount++] = 0xA0;
+	else if (tms_and_tdi == (TMS_1 | TDI_0))
+		pucBuffer[ucCount++] = 0xA2;
+	else if (tms_and_tdi == (TMS_1 | TDI_1))
+		pucBuffer[ucCount++] = 0xA2;
+	pucBuffer[ucCount++] = 0xEB;
+
+	mpsse_write_read(local_ctx, pucBuffer, ucCount, NULL, 0);
+}
+
+/**
+* Sends a dummy packet to the device
+*/
+static void dummy_scan_packet(void)
+{
+	transitions("0",TDI_0);
+}
+
+/**
+* Sends a check packet to the device
+*/
+static void send_check_packet(void)
+{
+	unsigned char pucBuffer[3];
+	unsigned char ucCount = 0;
+
+	if (in_normal_scan_mode())
+		pucBuffer[ucCount++] = 0x4B;
+	else
+		pucBuffer[ucCount++] = 0x1B;
+	pucBuffer[ucCount++] = 0x03;
+	pucBuffer[ucCount++] = 0x00;
+
+	mpsse_write_read(local_ctx, pucBuffer, ucCount, NULL, 0);
+}
+
+/**
+*  Sends a check packet to the device
+*/
+static void check_packet(void)
+{
+	/* dummy RTI->RTI when in oscan1. */
+	bool in_oscan1 = !in_normal_scan_mode();
+	if (in_oscan1)
+	{
+		dummy_scan_packet();
+		/* From pranab:
+		IEEE Std 1149.7-2009 
+		- Section "21.7: TAPC state/packet relationships" 
+		- with "Figure 21-9 - Typical DTS-initiated packet sequences and APU state transitions" 
+		contains the picture I created earlier for Check packets & Dummy scan packets.
+		*/
+	}
+	send_check_packet();
+}
+
+/**
+*  Enters oscan1 mode.
+*/
+static void enter_oscan1(void)
+{
+	/* Enter the command STFMT MODE
+	   where :
+	   parameter STFMT    = 5'h03;  Store Format
+	   parameter TAP7_OSCAN1 = 5'h9; */
+	send_2part_command(STFMT,OSCAN1);
+	/* After the command, must send a check packet */
+	check_packet();
+	/* Now in oscan1.  Port must know this. */
+	set_format(oscan1);
+}
+
+/**
+* Enters 2 wire_mode
+* @return 1: Success, 0 : Failure
+*/
+static int enter_2wire_mode_and_test(void)
+{
+	int wr = want_result(0);
+
+	set_JTAG_to_idle();
+
+	/* Ahead of discovery, enter 2wire mode, and discover in that mode.
+	   All JTAGs should be set to idle at this point. */
+	enter_control_mode(2);
+
+	/* cl 2 should now be locked and the arcs disconnected.
+	   The 2wire device has now disconnected the ARCs. */
+	enter_oscan1();
+
+	/* Get out of disconnected state by going to IR and RTI
+	   Now reconnect the arcs by entering shift-IR and then back to
+	   RTI. */
+	transitions("1100110",TDI_0);
+
+	/* Now we can continue discovering the ARCs. */
+	want_result(wr);
+	return 1;
+}
+
+/**
+* Checks if we are in normal / oscan1 mode.
+* @return TRUE : Normal scan mode, FALSE : oscan1 mode.
+*/
+static bool in_normal_scan_mode(void)
+{
+	return tycJtagScanFormat == normal;
+}
+
+/**
+* Sets the current scan mode.
+* @param sf contains the Scan mode format
+*/
+static void set_format(TycJtagScanFormat sf)
+{
+	unsigned char normalFormatSeq[] = { 0x80,0xA0,0xEB,0x82,
```

Comment:
Same formatting problem as `pucInitializeSequence`. I would format this as having three columns: command, data, direction, this will make it easier to read, especially since all commands here follow this pattern.
```c
	unsigned char normalFormatSeq[] = {
		0x80,0xA0,0xEB,
		0x82,0x60,0x60,
		0x80,0xA0,0xEB,
		0x82,0x00,0x60,
		0x80,0xE0,0xEB,
		0x82,0x00,0x60
	};
```